### PR TITLE
Chart refactor

### PR DESCRIPTION
**Pull request by FedeBev**
Since there's still an [open issue](https://github.com/distribworks/dkron/issues/676) in the main repo and I've the feeling this chart has been dropped, I've decided to rework it completely.

The main differences are:
- different values for server and agent
- HPA
- it works (seriously, the original didn't work for me because of [this issue](https://github.com/distribworks/dkron/issues/676#issuecomment-586953312))
- server is deployed using a StatefulSet, allowing the usage of PV (through volume claim template), resolving #2
- a custom image can be specified, useful when an image with a custom plugin should be used
- I've no longer seen #1, maybe just because of dkron upgrade (3.1.6)

Known issues:
- no health check specified, always [because of this](https://github.com/distribworks/dkron/issues/676#issuecomment-586953312) (help wanted)

Let me know what are your thoughts, I'm open to feedback available to make any changes

### CODE REVIEW

**Comment by victorcoder** [comment]
Sorry for the mega-delay here @FedeBev, thank you very much for the work! I had not prioritized the K8s work until now, going to test this and merge if it works.

File: dkron/templates/envsConfigmap.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "dkron.fullname" . }}-envs
+  labels:
+    {{- include "dkron.labels" . | nindent 4 }}
+data:
+  {{- range $key, $value := .Values.envs }}
+    {{ $key }}: {{ $value | quote }}
+  {{- end }}
```
Comment (victorcoder) [inline dkron/templates/envsConfigmap.yaml:10]: Line feed

File: dkron/templates/serviceaccount.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    {{- include "dkron.labels" . | nindent 4 }}
+  {{- with .Values.serviceAccount.annotations }}
+  annotations:
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
+{{- end }}
+{{ if .Values.image.pullSecretName -}}
+imagePullSecrets:
+- name: {{ .Values.image.pullSecretName }}
+{{- end }}
```
Comment (victorcoder) [inline dkron/templates/serviceaccount.yaml:16]: LF

**Review by victorcoder — CHANGES_REQUESTED** [review]
Aside from the nitpicks, it LGTM but I have been unable to do a rolling restart without breaking the cluster, what's your strategy?

**Comment by FedeBev** [comment]
What kind of problem are you facing @victorcoder? I just tried a rolling update from version v3.1.8 to v3.1.10 and everything went good except for a slight leader election delay.

Do you mind sharing your values and some details about your cluster?

I've just fixed the missing newlines and cleaned up the sample values file

**Comment by victorcoder** [comment]
I just do a `helm install` `helm upgrade` and the rolling is too fast for the leader election to happen correctly, when done rolling there's no leader in the cluster, all nodes ends up as followers.

Are you tweeking some other options?

**Comment by FedeBev** [comment]
No additional tweek on my configuration.

This also happens to me, but after a few seconds after the end of the rollout a new leader is elected again and everything goes fine.

We can try to tune the rollout strategy, but taking into account the [issue about the kubernetes discovery](https://github.com/distribworks/dkron/issues/676#issuecomment-586953312), I don't think we can achieve something better with helm. An operator would solve the issue, but it's a huge effort.

**Comment by vcastellm** [comment]
@FedeBev what about a liveness, readiness check?

**Comment by FedeBev** [comment]
@vcastellm in the  [issue about the kubernetes discovery](https://github.com/distribworks/dkron/issues/676#issuecomment-586953312) I've mentioned before, you can find why I wasn't able to use those.

I don't see any other possible way right now. I guess we should change something about how dkron starts or request a PR to the kubernetes discovery library.
In my opinion, the point is that dkron should be able to discover itself when the pod is running (/health returns 200), NOT when is ready. This way the k8s discovery works and we can implement a useful /health and ready probe.

**Comment by vcastellm** [comment]
Understood, thanks, I will take a look on how we can improve the /health endpoint

**Comment by FedeBev** [comment]
Keep me posted, I'd be glad to help.
Dkron could become a core component for a project I'm working on  however our platform relies entirely on K8s and this is a big issue

**Comment by Espina2** [comment]
I'm also very interested in this fix. @vcastellm @FedeBev anything that I can do to make this happen? Also, is this not better than the currently available chart?

**Comment by FedeBev** [comment]
@Espina2 the chart is working on my cluster in a dev environment, but it's still very young and there's still a long way to go before being production ready.
There's an issue with the k8s discovery that makes you dkron cluster unavailable due to master election during the rolling upgrade.

**Comment by omarzouk** [comment]
hello! also highly interested in this. Has anyone come up with a way to work around the discovery issue? any thoughts on how to proceed?

**Comment by vcastellm** [comment]
Merged this PR and added some changes, I'll take over from here.

**Pull request merged**